examples/tpm2-extendpcr/extendpcr.go

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2018, Ian Haken. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Extends a pcr with some data.
+package main
+
+import (
+	"encoding/hex"
+	"flag"
+	"fmt"
+	"os"
+
+	"github.com/google/go-tpm/tpm2"
+	"github.com/google/go-tpm/tpmutil"
+)
+
+var (
+	tpmPath = flag.String(
+		"tpm-path",
+		"/dev/tpm0",
+		"Path to the TPM device (character device or a Unix socket).")
+	pcr = flag.Int(
+		"pcr",
+		-1,
+		"Pcr to read. Must be within [0, 23].")
+	data = flag.String(
+		"data",
+		"",
+		"The hex encoded bytes with which to extend the PCR. Must not exceed 1024 bytes.")
+)
+
+func main() {
+	flag.Parse()
+	var errors []error
+	seal(&errors)
+	for _, err := range errors {
+		fmt.Fprintf(os.Stderr, "Error: %s\n", err)
+	}
+	if len(errors) > 0 {
+		os.Exit(1)
+	}
+}
+
+func seal(errors *[]error) {
```

Comment:
return a single error instead.
For `rwc.Close()` error, ignore it if a non-nil error is already being returned.

Reply:
The issue here is that rwc.Close is called by a deferred function, which needs some way to return errors.

Secondly, if resources aren't cleaned up (if Close fails) and this is not reported it could lead to a mysterious case of resource exhausted type errors. One concrete example that is similar and is present in example commands that I'll be submitting for review shortly is flushing objects, which if fail can cause TPM_RC_OBJECT_MEMORY. So I think it's important to report these errors regardless if there is also an additional error.

Reply:
In this case, though, this is a binary. The system will close the file handle when the binary exits, even if an error happens. I agree that this would be necessary in library code or for explicit handles in the TPM, but why for rwc?

Reply:
Agreed -- I think there are multiple valid perspectives on this. The one that appeals to me is that if a method returns an error, then report it. Making exceptions for special cases leaves room for unexpected situations.

Suppose someone changes the implementation of OpenTPM so that closing the object does more than just release file descriptors...after all, https://golang.org/pkg/io/#Closer is generic. Perhaps this is a bit contrived, but ignoring error handling to save a few lines seems like the wrong tradeoff.

Reply:
How about fmt printing the error in the deferred function then? Since this is not a library function, there's no need to use an array of errors to pass back an extra error that will immediately be printed anyway. Then this reports the error, and the rest of the error handling can return a single error, as originally suggested.

Reply:
Then the program will return success even if Close fails?